examples/t2i_adapter/train_t2i_adapter.py

```diff
@@ -0,0 +1,1454 @@
+                    latents = latents.to(weight_dtype)
+
+                if args.control_type == "depth":
+                    if args.depth_model_name_or_path == "zoe":
+                        with torch.autocast("cuda", enabled=True):
+                            depth_map = depth_model.infer(control_image)
+
+                        control_image = [
+                            torch.tensor(process_zoe_depth_map(depth_map[i])) for i in range(depth_map.shape[0])
+                        ]
+                        control_image = torch.stack(control_image, dim=0).to(accelerator.device, dtype=weight_dtype)
+                        control_image = control_image.unsqueeze(1)
+                    else:
+                        control_image = control_image.to(weight_dtype)
+                        with torch.autocast("cuda"):
+                            depth_map = depth_model(control_image).predicted_depth
+                        depth_map = torch.nn.functional.interpolate(
+                            depth_map.unsqueeze(1),
+                            size=image.shape[2:],
+                            mode="bicubic",
+                            align_corners=False,
+                        )
+                        depth_min = torch.amin(depth_map, dim=[1, 2, 3], keepdim=True)
+                        depth_max = torch.amax(depth_map, dim=[1, 2, 3], keepdim=True)
+                        depth_map = (depth_map - depth_min) / (depth_max - depth_min)
+
+                    control_image = (depth_map * 255.0).to(torch.uint8).float() / 255.0  # hack to match inference
+                    control_image = torch.cat([control_image] * 3, dim=1)
+                elif args.control_type == "sketch":
+                    edge_map = sketch_model(control_image)[-1]
+                    control_image = random_threshold(edge_map).to(torch.float32)
+                    # if detection resolution is specified, resize the control image back to the original resolution
+                    if args.detection_resolution is not None:
+                        control_image = torch.nn.functional.interpolate(
+                            control_image,
+                            size=(args.resolution, args.resolution),
+                            mode="bicubic",
+                            align_corners=False,
+                        )
+                    control_image = torch.cat([control_image] * 3, dim=1)
+                elif args.control_type == "lineart":
+                    line_map = lineart_model(control_image, resize_to=(args.resolution, args.resolution))
+                    control_image = line_map.float() / 255.0
+                    control_image = torch.cat([control_image] * 3, dim=1)
+
+                # Sample noise that we'll add to the latents
+                noise = torch.randn_like(latents)
+                bsz = latents.shape[0]
+
+                # Cubic sampling to sample a random timestep for each image
```

Comment:
Where does this trick come from?

Reply:
T2IAdapter paper, section 3.5.

The rationale behind this is that most of the content in the image is formed at the initial and mid-timesteps, and it's refined in later steps. And since we don't inject timestep information in T2IAdapter, during training, we sample timesteps for the initial range to increase its effect there.